daml-script/daml/Daml/Script.daml

```diff
@@ -566,6 +543,81 @@ instance Show Exercised where
     showsPrec 0 childEvents .
     showString "}"
 
+-- | HIDE This is an early access feature.
+data TreeIndex t
+  = CreatedIndex (CreatedIndexPayload t)
+  | ExercisedIndex (ExercisedIndexPayload t)
+
+-- | HIDE This is an early access feature.
+data CreatedIndexPayload t = CreatedIndexPayload
+  with
+    templateId : TemplateTypeRep
+    offset : Int
+
+-- | HIDE This is an early access feature.
+data ExercisedIndexPayload t = ExercisedIndexPayload
+  with
+    templateId : TemplateTypeRep
+    choice : Text
+    offset : Int
+    child : TreeIndex t
+
+-- | HIDE This is an early access feature.
+fromTree : Template t => TransactionTree -> TreeIndex t -> ContractId t
+fromTree tree index = fromTreeGo index tree.rootEvents
+
+fromTreeGo : Template t => TreeIndex t -> [TreeEvent] -> ContractId t
+fromTreeGo (CreatedIndex index) events =
+  case mapOptional fromCreated events of
```

Comment:
Probably perfectly fine for this PR but having just spent some time on optimizations: This type of code is usually much faster in Daml if you hand-write the selector since there is no list fusion so you pay for the full `mapOptional` and the allocation of the list + the second traversal for the index.